Test/gsapTests/EventDrivenPrognoserTests.cpp

```diff
@@ -0,0 +1,212 @@
+            double dataValue = datapt.second.get();
+            time_point timept = time_point(milliseconds(datapt.second.getTime()));
+            time = timept > time ? timept : time;
+            ScalarMessage<double>* message =
+                new ScalarMessage<double>(datapt.first, SENSOR_SRC, timept, dataValue);
+            bus.publish(std::shared_ptr<Message>(message));
+        }
+    }
+
+    ProgEvent publish(std::map<MessageId, Datum<double>> data) {
+        Require(data.size() > 0, "Must include atleast one data point to publish");
+        using std::chrono::milliseconds;
+        time_point time = time_point::min();
+
+        // Setup Messages
+        std::vector<ScalarMessage<double>*> messages;
+        for (auto&& datapt : data) {
+            double dataValue = datapt.second.get();
+            time_point timept = time_point(milliseconds(datapt.second.getTime()));
+            time = timept > time ? timept : time;
+            messages.push_back(
+                new ScalarMessage<double>(datapt.first, SENSOR_SRC, timept, dataValue));
+        }
+
+        // Setup request
+        Request request(time);
+        unique_lock requests_lock(processMessageMut);
+        requests.push(&request);
+        // Publish Message
+        for (auto message : messages) {
+            bus.publish(std::shared_ptr<Message>(message));
+        }
+        requests_lock.unlock();
+
+        unique_lock lock(request.mut);
+        request.waiting = true;
+        while (request.waiting) {
+            request.condition.wait(lock);
+        }
+
+        // Process Result
+        ProgEvent result = *currentResult;
+        delete currentResult; // Make sure memory is freed
+        return result;
+    }
+
+    /**
+     * Handles messages representing updates to the model inputs and
+     * outputs. When sufficient new data is collected, automaticlly triggers
+     * an observer step and publishes the result.
+     *
+     * @param message. The message to process.
+     **/
+    void processMessage(const std::shared_ptr<Message>& rawMessage) override {
+        switch (rawMessage->getMessageId()) {
+        case PCOE::MessageId::ModelStateVector:
+            printf("Test\n");
+            break;
+            {
+            case PCOE::MessageId::TestEvent0:
+                lock_guard guard(processMessageMut); // Process one ProgEvent at once
+                ProgEventMessage* message = dynamic_cast<ProgEventMessage*>(rawMessage.get());
+                ProgEvent result = message->getValue();
+                currentResult = new ProgEvent(result);
+
+                while (requests.size() > 0 && requests.front()->time <= message->getTimestamp()) {
+                    Request* currentRequest = requests.front();
+                    requests.pop();
+
+                    unique_lock lock(currentRequest->mut);
+                    currentRequest->waiting = false;
+                    currentRequest->condition.notify_one();
```

Comment:
Call `lock.unlock` before notify so that the notified thread doesn't wake up and immediately block when trying to re-aquire the mutex.